jena-text/src/main/java/org/apache/jena/query/text/assembler/TextIndexLuceneAssembler.java

```diff
@@ -61,7 +74,12 @@ public TextIndex open(Assembler a, Resource root, Mode mode) {
             if ( n.isLiteral() ) {
                 String literalValue = n.asLiteral().getLexicalForm() ; 
                 if (literalValue.equals("mem")) {
-                    directory = new RAMDirectory() ;
+                	// on RAMDirectory per node, not one per invocation
+                	directory = ramDirectories.get(root);
+                	if (directory == null) {
+                        directory = new RAMDirectory() ;
+                        ramDirectories.put(root, (RAMDirectory) directory);
```

Comment:
Changing `Directory directory;` above to `RAMDirectory directory;` would remove the cast.